File: database/import_sackmann.py
"""
import_sackmann.py
=================
Importa i dati del dataset Jeff Sackmann / Tennis Abstract nel database Supabase.

Attribuzione obbligatoria:
  Jeff Sackmann / Tennis Abstract (CC-BY-NC-SA 4.0)
  https://github.com/JeffSackmann/tennis_atp

Requisiti:
  pip install pandas supabase python-dotenv

Configurazione:
  Copia .env.local e aggiungi SUPABASE_SERVICE_KEY
  (usa la service_role key, non la anon key)
"""

import os
import re
import sys
import pandas as pd
from supabase import create_client, Client
from dotenv import load_dotenv

load_dotenv('.env.local')

SUPABASE_URL = os.environ.get('NEXT_PUBLIC_SUPABASE_URL', '')
SUPABASE_KEY = os.environ.get('SUPABASE_SERVICE_KEY', '')  # service_role key

if not SUPABASE_URL or not SUPABASE_KEY:
    print('ERROR: configura NEXT_PUBLIC_SUPABASE_URL e SUPABASE_SERVICE_KEY in .env.local')
    sys.exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

# Anni da importare (Sprint 1: prime stagioni iconiche)
YEARS = range(1993, 1998)  # 1993-1997

# Mapping surface Sackmann → nostro schema
SURFACE_MAP = {
    'Hard':   'Hard',
    'Clay':   'Clay',
    'Grass':  'Grass',
    'Carpet': 'Carpet',
    'I':      'Hard',   # Indoor hard
}

# Mapping round Sackmann → nostro schema
ROUND_MAP = {
    'F':   'F',
    'SF':  'SF',
    'QF':  'QF',
    'R16': 'R16',
    'R32': 'R32',
    'R64': 'R64',
    'R128':'R128',
}

# Tornei GrandSlam nel dataset Sackmann
SLAM_TOURNAMENTS = {
    'Australian Open': 'australian-open',
    'Roland Garros':   'roland-garros',
    'Wimbledon':       'wimbledon',
    'US Open':         'us-open',
}


def slugify(text: str) -> str:
    text = text.lower().strip()
    text = re.sub(r'[àáâãäå]', 'a', text)
    text = re.sub(r'[èéêë]',   'e', text)
    text = re.sub(r'[ìíîï]',   'i', text)
    text = re.sub(r'[òóôõö]',  'o', text)
    text = re.sub(r'[ùúûü]',   'u', text)
    text = re.sub(r'[ý]',      'y', text)
    text = re.sub(r'[ñ]',      'n', text)
    text = re.sub(r'[ç]',      'c', text)
    text = re.sub(r'[^a-z0-9\s-]', '', text)
    text = re.sub(r'[\s]+', '-', text)
    text = re.sub(r'-+', '-', text)
    return text.strip('-')


def get_or_create_player(row, prefix: str) -> int | None:
    """Crea il giocatore se non esiste e restituisce l'id."""
    first  = str(row.get(f'{prefix}_fname', '') or '').strip()
    last   = str(row.get(f'{prefix}_lname', '') or '').strip()
    ioc    = str(row.get(f'{prefix}_ioc',   '') or '').strip()
    hand   = str(row.get(f'{prefix}_hand',  '') or '').strip()
    ht     = row.get(f'{prefix}_ht')
    dob    = row.get(f'{prefix}_dob')
    sid    = row.get(f'{prefix}_id')

    if not first and not last:
        return None

    slug = slugify(f'{first}-{last}')

    # Cerca per sackmann_id
    res = supabase.table('players').select('id').eq('sackmann_id', int(sid) if pd.notna(sid) else -1).execute()
    if res.data:
        return res.data[0]['id']

    # Cerca per slug
    res = supabase.table('players').select('id').eq('slug', slug).execute()
    if res.data:
        return res.data[0]['id']

    # Crea nuovo
    data = {
        'sackmann_id': int(sid) if pd.notna(sid) else None,
        'slug':        slug,
        'first_name':  first,
        'last_name':   last,
        'country_code': ioc[:3] if ioc else None,
        'hand':        hand[:1].upper() if hand and hand[:1].upper() in ('R', 'L') else None,
        'height_cm':   int(ht) if pd.notna(ht) else None,
        'birth_date':  str(dob)[:10] if pd.notna(dob) else None,
    }
    res = supabase.table('players').insert(data).execute()
    return res.data[0]['id'] if res.data else None


def get_tournament_id(tourney_name: str) -> int | None:
    slug = SLAM_TOURNAMENTS.get(tourney_name)
    if not slug:
        return None
    res = supabase.table('tournaments').select('id').eq('slug', slug).execute()
    return res.data[0]['id'] if res.data else None


def import_year(year: int) -> None:
    url = f'https://raw.githubusercontent.com/JeffSackmann/tennis_atp/master/atp_matches_{year}.csv'
    print(f'\n→ Scaricando {year}...', end=' ', flush=True)

    try:
        df = pd.read_csv(url)
    except Exception as e:
        print(f'ERRORE: {e}')
        return

    # Filtra solo GrandSlam (fase 1)
    slams = df[df['tourney_name'].isin(SLAM_TOURNAMENTS.keys())]
    print(f'{len(slams)} partite GrandSlam trovate')

    imported = 0
    errors   = 0

    for _, row in slams.iterrows():
        try:
            winner_id = get_or_create_player(row, 'winner')
            loser_id  = get_or_create_player(row, 'loser')
            tourn_id  = get_tournament_id(str(row.get('tourney_name', '')))

            if not winner_id or not loser_id:
                continue

            # Costruisce lo slug
            wn  = slugify(f"{row.get('winner_fname','')} {row.get('winner_lname','')}")
            ln  = slugify(f"{row.get('loser_fname', '')} {row.get('loser_lname', '')}")
            tr  = slugify(str(row.get('tourney_name', '')))
            rnd = str(row.get('round', '')).lower().replace(' ', '-')
            slug = f'{wn}-vs-{ln}-{tr}-{year}-{rnd}'

            match_data = {
                'sackmann_id':  str(row.get('match_num', '')) or None,
                'slug':         slug,
                'tournament_id': tourn_id,
                'year':         year,
                'match_date':   str(row.get('tourney_date', ''))[:10] if pd.notna(row.get('tourney_date')) else None,
                'round':        ROUND_MAP.get(str(row.get('round', '')), str(row.get('round', ''))[:4]),
                'surface':      SURFACE_MAP.get(str(row.get('surface', '')), 'Hard'),
                'winner_id':    winner_id,
                'loser_id':     loser_id,
                'winner_rank':  int(row['winner_rank']) if pd.notna(row.get('winner_rank')) else None,
                'loser_rank':   int(row['loser_rank'])  if pd.notna(row.get('loser_rank'))  else None,
                'score':        str(row.get('score', '')),
                'duration_min': int(row['minutes']) if pd.notna(row.get('minutes')) else None,

                'w_ace':    _int(row.get('w_ace')),
                'w_df':     _int(row.get('w_df')),
                'w_svpt':   _int(row.get('w_svpt')),
                'w_1stIn':  _int(row.get('w_1stIn')),
                'w_1stWon': _int(row.get('w_1stWon')),
                'w_2ndWon': _int(row.get('w_2ndWon')),
                'w_SvGms':  _int(row.get('w_SvGms')),
                'w_bpSaved':_int(row.get('w_bpSaved')),
                'w_bpFaced':_int(row.get('w_bpFaced')),

                'l_ace':    _int(row.get('l_ace')),
                'l_df':     _int(row.get('l_df')),
                'l_svpt':   _int(row.get('l_svpt')),
                'l_1stIn':  _int(row.get('l_1stIn')),
                'l_1stWon': _int(row.get('l_1stWon')),
                'l_2ndWon': _int(row.get('l_2ndWon')),
                'l_SvGms':  _int(row.get('l_SvGms')),
                'l_bpSaved':_int(row.get('l_bpSaved')),
                'l_bpFaced':_int(row.get('l_bpFaced')),
            }

            supabase.table('matches').upsert(match_data, on_conflict='slug').execute()
            imported += 1

        except Exception as e:
            errors += 1
            if errors < 5:
                print(f'\n  WARN: {e}')

    print(f'  ✓ {imported} importate, {errors} errori')


def _int(val) -> int | None:
    try:
        return int(val) if pd.notna(val) else None
    except (ValueError, TypeError):
        return None


if __name__ == '__main__':
    print('Ace Chronicle — Import Sackmann')
    print('Attribuzione: Jeff Sackmann / Tennis Abstract (CC-BY-NC-SA 4.0)')
    print('https://github.com/JeffSackmann/tennis_atp\n')

    for year in YEARS:
        import_year(year)

    print('\n✅ Import completato.')

File: database/import_sackmann.mjs
/**
 * Importa tutti i Grandi Slam ATP dal dataset Jeff Sackmann (GitHub)
 * Uso: node database/import_sackmann.mjs
 * Dopo: node database/migrate_cultural_impact.mjs
 */
import postgres from 'postgres'

const db = postgres(process.env.DATABASE_URL, { max: 1 })
const BASE = 'https://raw.githubusercontent.com/JeffSackmann/tennis_atp/master'
const YEAR_FROM = 1980
const YEAR_TO   = 2002

// ── CSV helpers ───────────────────────────────────────────────────────────────

function splitLine(line) {
  const out = []; let cur = ''; let inQ = false
  for (const c of line) {
    if (c === '"') inQ = !inQ
    else if (c === ',' && !inQ) { out.push(cur); cur = '' }
    else cur += c
  }
  out.push(cur)
  return out
}

function parseCSV(text) {
  const lines = text.trim().split('\n').map(l => l.replace(/\r$/, ''))
  const headers = splitLine(lines[0]).map(h => h.trim())
  return lines.slice(1).filter(Boolean).map(line => {
    const vals = splitLine(line)
    return Object.fromEntries(headers.map((h, i) => [h, (vals[i] ?? '').trim()]))
  })
}

function toSlug(str) {
  return (str || '')
    .normalize('NFD').replace(/[̀-ͯ]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

function num(v) { const n = parseInt(v); return isNaN(n) ? null : n }

function toDate(yyyymmdd) {
  if (!yyyymmdd || yyyymmdd.length < 8) return null
  const s = yyyymmdd.replace(/\D/g, '')
  return `${s.slice(0,4)}-${s.slice(4,6)}-${s.slice(6,8)}`
}

const SLAM_SLUGS = {
  'australian open': 'australian-open',
  'roland garros':   'roland-garros',
  'wimbledon':       'wimbledon',
  'us open':         'us-open',
}

function slamSlug(name) {
  const n = (name || '').toLowerCase()
  for (const [k, v] of Object.entries(SLAM_SLUGS)) if (n.includes(k)) return v
  return null
}

// ── 1. Scarica e filtra partite per anno ──────────────────────────────────────

console.log(`⏳  Download Grandi Slam ${YEAR_FROM}–${YEAR_TO}…\n`)

const allRows   = []
const neededIds = new Set()

for (let yr = YEAR_FROM; yr <= YEAR_TO; yr++) {
  const url = `${BASE}/atp_matches_${yr}.csv`
  try {
    const res = await fetch(url)
    if (!res.ok) { console.warn(`  ⚠  ${yr}: HTTP ${res.status}`); continue }
    const rows = parseCSV(await res.text()).filter(r =>
      r.tourney_level === 'G' && r.score?.trim() && r.winner_id && r.loser_id
    )
    rows.forEach(r => { neededIds.add(r.winner_id); neededIds.add(r.loser_id) })
    allRows.push(...rows)
    console.log(`  ✓  ${yr}  ${String(rows.length).padStart(4)} partite Slam`)
  } catch (e) { console.warn(`  ⚠  ${yr}: ${e.message}`) }
}

console.log(`\n  Totale: ${allRows.length} partite · ${neededIds.size} giocatori unici\n`)

// ── 2. Scarica giocatori ──────────────────────────────────────────────────────

console.log('⏳  Download atp_players.csv…')
const pRes  = await fetch(`${BASE}/atp_players.csv`)
const pRows = parseCSV(await pRes.text()).filter(p => neededIds.has(p.player_id))
console.log(`  ✓  ${pRows.length} giocatori da importare\n`)

// ── 3. Pulizia matches ────────────────────────────────────────────────────────

console.log('🗑   Pulizia partite…')
await db`TRUNCATE matches RESTART IDENTITY CASCADE`
console.log('  ✓  Partite svuotate\n')

// ── 4. Inserisci giocatori ────────────────────────────────────────────────────

console.log('⏳  Inserimento giocatori…')

// Sackmann CSV usa "first_name"/"last_name" o "name_first"/"name_last" a seconda della versione
const getFirst = p => (p.first_name || p.name_first || '').trim()
const getLast  = p => (p.last_name  || p.name_last  || '').trim()

let pInserted = 0
for (const p of pRows) {
  const first = getFirst(p)
  const last  = getLast(p)
  if (!first || !last) continue
  const slug = `${toSlug(first)}-${toSlug(last)}`
  if (!slug || slug === '-') continue

  try {
    await db`
      INSERT INTO players (sackmann_id, slug, first_name, last_name, country_code, hand, birth_date, height_cm, grand_slams)
      VALUES (
        ${num(p.player_id)}, ${slug}, ${first}, ${last},
        ${p.country || p.ioc || null},
        ${p.hand === 'R' || p.hand === 'L' ? p.hand : null},
        ${toDate(p.dob)}, ${num(p.height)}, 0
      )
      ON CONFLICT (slug) DO UPDATE SET
        sackmann_id  = COALESCE(players.sackmann_id, EXCLUDED.sackmann_id),
        country_code = COALESCE(players.country_code, EXCLUDED.country_code),
        hand         = COALESCE(players.hand,         EXCLUDED.hand),
        birth_date   = COALESCE(players.birth_date,   EXCLUDED.birth_date),
        height_cm    = COALESCE(players.height_cm,    EXCLUDED.height_cm)
    `
    pInserted++
  } catch {
    // Slug collision con sackmann_id diverso → aggiunge suffisso numerico
    try {
      await db`
        INSERT INTO players (sackmann_id, slug, first_name, last_name, country_code, hand, birth_date, height_cm, grand_slams)
        VALUES (${num(p.player_id)}, ${slug + '-' + p.player_id}, ${first}, ${last},
                ${p.country || null}, ${p.hand === 'R' || p.hand === 'L' ? p.hand : null},
                ${toDate(p.dob)}, ${num(p.height)}, 0)
        ON CONFLICT (sackmann_id) DO NOTHING
      `
      pInserted++
    } catch { /* skip */ }
  }
}

console.log(`  ✓  ${pInserted} giocatori inseriti/aggiornati\n`)

// ── 5. Mappe ID ───────────────────────────────────────────────────────────────

const dbPlayers = await db`SELECT id, sackmann_id, slug FROM players WHERE sackmann_id IS NOT NULL`
const pidMap    = new Map(dbPlayers.map(r => [String(r.sackmann_id), r])) // sackmannId → {id, slug}

const dbTourns  = await db`SELECT id, slug FROM tournaments`
const tournMap  = new Map(dbTourns.map(r => [r.slug, r.id]))

// ── 6. Inserisci partite ──────────────────────────────────────────────────────

console.log('⏳  Inserimento partite…')
let mInserted = 0, mSkipped = 0

for (const r of allRows) {
  const tSlug = slamSlug(r.tourney_name)
  const tId   = tSlug ? tournMap.get(tSlug) : null
  const winner = pidMap.get(r.winner_id)
  const loser  = pidMap.get(r.loser_id)
  const round  = r.round?.trim()
  const year   = num(r.tourney_date?.slice(0, 4))

  if (!tId || !winner || !loser || !round || !year) { mSkipped++; continue }

  const slug = `${winner.slug}-vs-${loser.slug}-${tSlug}-${year}-${round.toLowerCase()}`

  try {
    await db`
      INSERT INTO matches (
        sackmann_id, slug, tournament_id, year, match_date, round, surface,
        winner_id, loser_id, winner_rank, loser_rank, score, duration_min,
        w_ace, w_df, w_svpt, w_1stIn, w_1stWon, w_2ndWon, w_SvGms, w_bpSaved, w_bpFaced,
        l_ace, l_df, l_svpt, l_1stIn, l_1stWon, l_2ndWon, l_SvGms, l_bpSaved, l_bpFaced,
        featured
      ) VALUES (
        ${r.tourney_id + '-' + r.match_num}, ${slug}, ${tId}, ${year},
        ${toDate(r.tourney_date)}, ${round}, ${r.surface || null},
        ${winner.id}, ${loser.id}, ${num(r.winner_rank)}, ${num(r.loser_rank)},
        ${r.score}, ${num(r.minutes)},
        ${num(r.w_ace)},  ${num(r.w_df)},    ${num(r.w_svpt)}, ${num(r.w_1stIn)},
        ${num(r.w_1stWon)}, ${num(r.w_2ndWon)}, ${num(r.w_SvGms)}, ${num(r.w_bpSaved)}, ${num(r.w_bpFaced)},
        ${num(r.l_ace)},  ${num(r.l_df)},    ${num(r.l_svpt)}, ${num(r.l_1stIn)},
        ${num(r.l_1stWon)}, ${num(r.l_2ndWon)}, ${num(r.l_SvGms)}, ${num(r.l_bpSaved)}, ${num(r.l_bpFaced)},
        false
      )
      ON CONFLICT (slug) DO NOTHING
    `
    mInserted++
  } catch { mSkipped++ }
}

console.log(`  ✓  ${mInserted} partite inserite · ${mSkipped} saltate\n`)

// ── 7. Calcola grand_slams e atp_peak_rank ────────────────────────────────────

console.log('⏳  Calcolo grand_slams e peak rank…')

await db`
  UPDATE players p SET grand_slams = (
    SELECT COUNT(*) FROM matches m
    JOIN tournaments t ON t.id = m.tournament_id
    WHERE m.winner_id = p.id AND m.round = 'F' AND t.category = 'GrandSlam'
  )
`

await db`
  UPDATE players p
  SET atp_peak_rank = sub.peak
  FROM (
    SELECT player_id, MIN(rank) AS peak FROM (
      SELECT winner_id AS player_id, winner_rank AS rank FROM matches WHERE winner_rank IS NOT NULL AND winner_rank > 0
      UNION ALL
      SELECT loser_id  AS player_id, loser_rank  AS rank FROM matches WHERE loser_rank  IS NOT NULL AND loser_rank  > 0
    ) r
    GROUP BY player_id
  ) sub
  WHERE sub.player_id = p.id
`

// ── Report finale ─────────────────────────────────────────────────────────────

const [{ n: totP }] = await db`SELECT COUNT(*) AS n FROM players`
const [{ n: totM }] = await db`SELECT COUNT(*) AS n FROM matches`
const [{ n: totY }] = await db`SELECT COUNT(DISTINCT year) AS n FROM matches`

console.log(`  ✓  grand_slams e peak_rank aggiornati\n`)
console.log(`✅  Import completato`)
console.log(`    ${totP} giocatori · ${totM} partite · ${totY} stagioni`)
console.log(`\n→  Esegui ora: node database/migrate_cultural_impact.mjs`)

await db.end()
